pkg/daemon/controller/controller.go

```diff
@@ -0,0 +1,324 @@
+	// TODO: Instead of reconciling services only once, we need to watch managed Windows services for changes and
+	//       reconcile services if a change happens, as well as reconciling if any changes occur to the ConfigMap
+	//       or node object.
+	return c.reconcileServices(data.Services)
+}
+
+// localIPv4Addresses returns a slice of all non-loopback ipv4 addresses
+func localIPv4Addresses() ([]net.IP, error) {
+	var ipv4Addresses []net.IP
+	netIfs, err := net.Interfaces()
+	if err != nil {
+		return nil, err
+	}
+	for _, netInterface := range netIfs {
+		addrs, err := netInterface.Addrs()
+		if err != nil {
+			return nil, err
+		}
+		for _, addr := range addrs {
+			ipAddr, ok := addr.(*net.IPNet)
+			if !ok {
+				continue
+			}
+			ipv4Addr := ipAddr.IP.To4()
+			if ipv4Addr != nil || ipv4Addr.IsLoopback() {
+				continue
+			}
+			ipv4Addresses = append(ipv4Addresses, ipv4Addr)
+		}
+	}
+	if len(ipv4Addresses) == 0 {
+		return nil, errors.New("no usable ipv4 addresses found")
+	}
+	return ipv4Addresses, nil
+}
+
+// currentNode returns the node associated with this VM
+func currentNode(nodes *core.NodeList, localAddrs []net.IP) (*core.Node, error) {
+	for _, localAddr := range localAddrs {
+		// Go through each node and check if the node has the ipv4 address in the address slice
+		for _, node := range nodes.Items {
+			for _, addr := range node.Status.Addresses {
+				if addr.Address == localAddr.String() {
+					return &node, nil
+				}
+			}
+		}
+	}
+	return nil, errors.New("unable to find associated node")
+}
+
+// NewServiceController returns a pointer to a ServiceController object
+func NewServiceController(client client.Client, mgr winsvc.Mgr, nodeName string) *ServiceController {
+	return &ServiceController{client: client, svcMgr: mgr, ctx: context.TODO(), nodeName: nodeName}
+}
+
+// existingServices returns a map with the keys being all Windows services currently present on the VM
+func (c *ServiceController) existingServices() (map[string]struct{}, error) {
+	// The most reliable way to determine if a service exists or not is to do a 'list' API call. It is possible to
+	// remove this call, and parse the error messages of a service 'open' API call, but I find that relying on human
+	// readable errors could cause issues when providing compatibility across different versions of Windows.
+	svcList, err := c.svcMgr.ListServices()
+	if err != nil {
+		return nil, err
+	}
+	svcs := make(map[string]struct{})
+	for _, service := range svcList {
+		svcs[service] = struct{}{}
+	}
+	return svcs, nil
+}
+
+// reconcileServices ensures that all the services passed in via the services slice are created, configured properly
+// and started
+func (c *ServiceController) reconcileServices(services []servicescm.Service) error {
+	// go through svcList
+	// sort services in ascending priority
+	sort.Slice(services, func(i, j int) bool {
+		return services[i].Priority < services[j].Priority
+	})
```

Comment:
I'm trying to think of a way to avoid doing this here, as the `servicescm.Parse()` call in `RunController()` already does a sort as part of validation. Maybe modifying [`validatePriorities()`](https://github.com/openshift/windows-machine-config-operator/blob/master/pkg/servicescm/servicescm.go#L255) to take in a pointer, *[]Service, so it actually modifies the argument

Reply:
Good point. Ideally, `services` should be in order. Is there any [Priority queue](https://en.wikipedia.org/wiki/Priority_queue) implementation we could use while creating the collection (`servicescm.Parse(...)`)? Alternatively, an [insertion-sort](https://en.wikipedia.org/wiki/Insertion_sort) approach may also be useful.

Reply:
Having looked into it i believe that because validatePriorities is called as part of Parse() it has already sorted the data so i can remove this code.
A slice is already a pointer so the backing arrays data will have been affected by the sort